Use interactive markers to control iiwa14 joints

sloretz opened the pull request:
Add Demo using interactive markers to control iiwa14 in drake.

![Peek 2020-10-19 13-54](https://user-images.githubusercontent.com/4175662/96510753-dcea8380-1212-11eb-89ca-4a9019a8a9cd.gif)

This uses pydrake systems framework, the manipulation station, and ROS 2 Rolling with ros/sdformat_urdf#1 to control the robot in joint space with interactive markers on each joint.

@iantheengineer FYI  - see interactive_demo.py


Roughly to run this:

1. Build a ROS 2 workspace with ros/sdformat_urdf#1 and source it in a terminal
2. Download the latest binary install of drake
3. Use @EricCousineau-TRI's trick of `python3 -m venv --system-site-packages path/to/extracted/drake` and activate the venv to make Drake usable from that terminal
4. In another terminal activate the venv and run `drake-visualizer`
5. In another terminal activate the venv and run `rviz2 -d path/to/this/repo/view.rviz`
6. Run `python3 ./interactive_demo.py`

sloretz committed:
Whatever is here - segfaults

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Marker controls robot-ish

Quickly gets stuck in singularity
Maybe add routine to re-home robot in joint space when Diff_ik fails?

Signed-off-by: Shane Loretz<sloretz@openrobotics.org>
Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Remove print statements

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Almost does what I want, diffik fails pretty quickly

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
At least it runs - does nothing useful but outputs to RViz again

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Add changes for joint space control - not sure what works

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Better view

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Almost does the right rotation of marker

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Fix bug where axis was affected by translation to F

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Faster update rate!

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Increment joint angle (still two joints go opposite of what I'd expect)

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Get rid of computed sdf

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Marker in child frame

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Behaves as expected

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Cleanup unused code

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Delete unused file

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Copy view.rviz -> interactive_demo.rviz

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Reset view.rviz for ros2_demo.py

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz commented:
@EricCousineau-TRI @azeey mind having a look?

EricCousineau-TRI commented:
Taking a look now!

EricCousineau-TRI commented on iiwa14_no_collision.sdf.in:
```diff
@@ -0,0 +1,469 @@
+        <geometry>
+          <mesh>
+            <scale>1 1 1</scale>
+            <uri>file://PWD_GOES_HERE/meshes/visual/link_3.obj</uri>
```
Is there some way to remove the need for template substituation?
Can we instead just hack ROS2's equivalent of a package path?

sloretz replied:
Maybe it could use `package://` to a ROS package with [`drake::multibody::PackageMap`](https://github.com/RobotLocomotion/drake/blob/afd584c4b38ca168d12ed864a079309ee1eed88c/multibody/parsing/package_map.h) being made aware of that package.

EricCousineau-TRI replied:
Yeah, it should already be aware.

FTR, we do want this to eventually be referred to `package://drake/...` so it won't step on the toes of other stuff.

sloretz replied:
I think it makes sense to continue this conversation on the other thread: https://github.com/sloretz/drake_ros2_demos/pull/3#discussion_r516817838 starting with  68615ac

EricCousineau-TRI commented on interactive_demo.py:
```diff
@@ -0,0 +1,224 @@
+from pydrake.common.value import AbstractValue
+from pydrake.examples.manipulation_station import ManipulationStation
+from pydrake.geometry import ConnectDrakeVisualizer
+from pydrake.manipulation.planner import DifferentialInverseKinematicsParameters
```
Can you format these according to this `isort` settings file?
https://github.com/RobotLocomotion/drake/pull/13709
https://github.com/RobotLocomotion/drake/blob/985cf2167fb5c388614137c65796772ad3e98f6d/pyproject.toml

sloretz replied:
isort'd in 95fb9c1

EricCousineau-TRI commented on interactive_demo.py:
```diff
@@ -0,0 +1,224 @@
+from pydrake.examples.manipulation_station import ManipulationStation
+from pydrake.geometry import ConnectDrakeVisualizer
+from pydrake.manipulation.planner import DifferentialInverseKinematicsParameters
+# from pydrake.manipulation.planner import DifferentialInverseKinematicsIntegrator
```
Can you run `flake8 --select=F` on these files?

sloretz replied:
Removed unused imports in 4641a7b

EricCousineau-TRI commented on interactive_demo.py:
```diff
@@ -0,0 +1,224 @@
+
+
+if __name__ == '__main__':
+    builder = DiagramBuilder()
```
This code is non-trivial, and should be in `main()` to avoid leaking out local variables.

Can you fix that here?

sloretz replied:
wrapped in `main()` in 0fb2f29

EricCousineau-TRI commented on iiwa14_no_collision.sdf.in:
```diff
@@ -0,0 +1,469 @@
+<?xml version="1.0"?>
```
Can you state why you need to vendor in this SDFormat file?

We have this in Drake:
https://github.com/RobotLocomotion/drake/blob/985cf2167fb5c388614137c65796772ad3e98f6d/manipulation/models/iiwa_description/sdf/iiwa14_no_collision.sdf

(albeit with bad package names)

sloretz replied:
The trouble is [the relative paths used for the meshes](https://github.com/RobotLocomotion/drake/blob/985cf2167fb5c388614137c65796772ad3e98f6d/manipulation/models/iiwa_description/sdf/iiwa14_no_collision.sdf#L29). When RViz loads the SDFormat robot description it only has a string, not the path it came from. I hacked around it by templating it to make `file://` URIs.

EricCousineau-TRI replied:
Yeah, understood... but can't that be solved by hacking the package path?

sloretz replied:
>  but can't that be solved by hacking the package path?

By `that` do you mean the vendoring or the relative URLs? Vendoring could be avoided by making an `ament_index` like layout in Drake. Relative URL's remain not useful on the ROS side.

Supporting `package://some_package_name` URIs in ROS 2 means creating a file `share/ament_index/resource_index/packages/some_package_name`, then putting all the relevant files into `share/some_package_name`. It also needs a stub `package.xml` in `share/some_package_name/package.xml` to get Drake's `PackageMap` to recognize it. See 68615ac which makes the ur10 demo demo do this.

```
share/
├── ament_index
│   └── resource_index
│       └── packages
│           └── ur10_description
└── ur10_description
    ├── meshes
    │   ├── base.mtl
    │   ├── base.obj
    │   ├── forearm.mtl
    │   ├── forearm.obj
    │   ├── shoulder.mtl
    │   ├── shoulder.obj
    │   ├── upperarm.mtl
    │   ├── upperarm.obj
    │   ├── wrist1.mtl
    │   ├── wrist1.obj
    │   ├── wrist2.mtl
    │   ├── wrist2.obj
    │   ├── wrist3.mtl
    │   └── wrist3.obj
    ├── package.xml
    └── ur10.sdf
```

I'll update the interactive demo to use `package://` URI's on the ROS side too, but the interactive demo is less interesting for showing how to work with Drake's package discovery mechanism because it's using the manipulation station diagram.

sloretz replied:
I forgot to mention this also means setting `AMENT_PREFIX_PATH` to get ROS to see the packages. Launching the UR10 demo now needs:

```
AMENT_PREFIX_PATH="$AMENT_PREFIX_PATH:$(pwd)" python3 ros2_demo.py
AMENT_PREFIX_PATH="$AMENT_PREFIX_PATH:$(pwd)" rviz2 -d view.rviz
```

sloretz replied:
Same treatment to interactive demo in 60e2013

EricCousineau-TRI replied:
Ahh, beautiful! Thank you!

FYI @traversaro Relating https://github.com/RobotLocomotion/drake-external-examples/issues/170

EricCousineau-TRI commented on drake_ros/drake_ros/systems/__init__.py:
```diff
@@ -1,10 +1,14 @@
 from .joint_state_publisher import JointStatePublisher
+from .moveable_joints import MoveableJoints
```
Er, is the spelling `Moveable` or `Movable`?

EricCousineau-TRI replied:
From https://github.com/RobotLocomotion/drake/commit/40876137bf2931f0a15641fc0f4a57fbd9c69c12:
```
$ grep -rnI -i 'moveable' | wc -l
12
$ grep -rnI -i 'movable' | wc -l
63
```

sloretz replied:
Moveable -> Movable in 357ad75

EricCousineau-TRI reviewed (commented):
Will run it on my machine

sloretz committed:
Removed unused imports

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Wrap in main()

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Use drake's isort settings

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Moveable -> Movable

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

azeey commented on drake_ros/drake_ros/systems/__init__.py:
```diff
@@ -1,10 +1,14 @@
 from .joint_state_publisher import JointStatePublisher
+from .moveable_joints import MovableJoints
```
`moveable` -> `movable`?

sloretz replied:
Oops. moveable -> movable in 6de6c7e534c0cef26b958fa024b9a7d36c69999b

azeey commented on interactive_demo.py:
```diff
@@ -0,0 +1,200 @@
+        'iiwa_joint_7']
+    joints = []
+    for name in joint_names:
+        joints.append(station.get_controller_plant().GetJointByName(name))
```
I'm getting the following error here:

```
Traceback (most recent call last):
  File "./interactive_demo.py", line 200, in <module>
    main()
  File "./interactive_demo.py", line 122, in main
    joints.append(station.get_controller_plant().GetJointByName(name))
TypeError: Unable to convert function return value to a Python type! The signature was
        (self: pydrake.examples.manipulation_station.ManipulationStation) -> drake::multibody::MultibodyPlant<double>
```

Any chance you know why?

azeey replied:
Found out why. I think `from pydrake.multibody.plant import MultibodyPlant` is needed in the imports to bring in the `drake::multibody::MultibodyPlant<double>` symbol

sloretz replied:
Seems like a bug in the python bindings for manipulation station. Worked around it by importing ManipulationStation in 29464598cf94a29c4efd386351b62522d560613a

azeey commented on interactive_demo.py:
```diff
@@ -0,0 +1,200 @@
+        fk_system.GetOutputPort('transforms'),
+        tf_system.GetInputPort('body_poses')
+    )
+    builder.Connect(
```
I don't see how `clock` is used in the `TFPublisher` system

sloretz replied:
`TFPublisher` gets the time from the clock input port and [sets the time of the transform from that](https://github.com/sloretz/drake_ros2_demos/blob/f70382b5c790b1aa2b7dc855fcfd70d4dbdeac0c/drake_ros/drake_ros/systems/tf_publisher.py#L67). This allows it to support both [system time](https://github.com/sloretz/drake_ros2_demos/blob/master/drake_ros/drake_ros/systems/system_clock.py) and [simulated time](https://github.com/sloretz/drake_ros2_demos/blob/master/drake_ros/drake_ros/systems/simulator_clock.py).

azeey replied:
Yup, missed it somehow.

azeey commented on drake_ros/drake_ros/systems/movable_joints.py:
```diff
@@ -0,0 +1,144 @@
+            (v[2], 0, -v[0]),
+            (-v[1], v[0], 0)), dtype=numpy.float64)
+        v_sub_x_squared = numpy.dot(v_sub_x, v_sub_x)
+        rotation_matrix = numpy.eye(3) + v_sub_x + v_sub_x_squared * (1.0 / (1.0 + c))
```
If `axis_hat` and `x_axis` are close to being anti-parallel, `c` will be close to -1.0 and cause a division by zero. I found a function in `pydrake.math` to compute this matrix.
```suggestion
        # What rotation would get the parent X axis to align with the joint axis?
        rotation_matrix = ComputeBasisFromAxis(0, axis_hat)
```
Feel free to ignore if this code is just for the iiwa.

azeey commented on drake_ros/drake_ros/systems/movable_point.py:
```diff
@@ -0,0 +1,117 @@
+
+
+
+class MovablePoint(LeafSystem):
```
I'm guessing this system is not needed anymore since we're not doing IK?

sloretz replied:
Yup. Removed unused `MovablePoint` system in 3860461

azeey commented on interactive_demo.py:
```diff
@@ -0,0 +1,200 @@
+    fk_system = builder.AddSystem(ForwardKinematics(station.get_controller_plant()))
+
+    # Blindly assume first joints in plant are the iiwa14
+    just_arm_joints = builder.AddSystem(
```
Is this (these two lines and `PadVector`) still needed?

sloretz replied:
Removed unused `PadVector` system 4b54925

azeey reviewed (approved):
LGTM! I was able to run this locally (in docker) after fixing a couple of issues I found (mentioned in comments). Given that this is a prototype, feel free to ignore some of my comments. As a general comment, I wonder how hard it would be to just publish joint states and use the Robot state publisher to do the TF computation instead of having the TFPublisher system here. This would further exercise the `sdformat_urdf` package

sloretz committed:
moveable -> movable

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Workaround manipulation station python binding bug

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Update drake_ros/drake_ros/systems/movable_joints.py

Co-authored-by: Addisu Z. Taddese <addisu@gmail.com>

sloretz committed:
Import ComputeBasisFromAxis

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Remove unused PadVector system

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Remove unused MovablePoint system

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Use package:// URIs for ur10 ros2 demo

Uses stub package.xml files and an extra ament index to convince both
ROS and Drake that the SDFormat XML files and meshes live in a ROS
package. This allows using `package://` URLs instead of the temporary
file made from populating a template with absolute `file://` URIs.

To launch:
AMENT_PREFIX_PATH="$AMENT_PREFIX_PATH:$(pwd)" python3 ros2_demo.py
AMENT_PREFIX_PATH="$AMENT_PREFIX_PATH:$(pwd)" rviz2 -d view.rviz

Signed-off-by: Shane Loretz <sloretz@openrobotics.org>
Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
os.path.join() instead of +

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Make Interactive demo use ament_index

Gets rid of templated sdformat file

To launch:
AMENT_PREFIX_PATH="$AMENT_PREFIX_PATH:$(pwd)" python3 interactive_demo.py
AMENT_PREFIX_PATH="$AMENT_PREFIX_PATH:$(pwd)" rviz2 -d interactive_demo.rviz

Signed-off-by: Shane Loretz <sloretz@openrobotics.org>
Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz commented:
@EricCousineau-TRI @azeey Feedback has been addressed. Mind taking another look?

EricCousineau-TRI reviewed (commented):
Can you use `.gitattributes` to suppress diffs for `.obj` files here?
https://github.com/RobotLocomotion/drake/blob/4731f31ad39753dfde25611456e55f98d027d469/.gitattributes

EricCousineau-TRI commented on interactive_demo.py:
```diff
@@ -0,0 +1,169 @@
+from pydrake.geometry import ConnectDrakeVisualizer
+from pydrake.math import RigidTransform
+from pydrake.multibody.parsing import PackageMap
+from pydrake.multibody.plant import MultibodyPlant
```
Ahhhh, shewt! I saw this commit about `ManipulationStation`.

Can you file a brief Drake issue? (I'll assign it to myself once you do)

sloretz replied:
Opened https://github.com/RobotLocomotion/drake/issues/14369

azeey reviewed (approved):
LGTM. Mind adding the `AMENT_PREFIX_PATH` stuff to the README as well?

sloretz committed:
Suppress diffs on obj files

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz commented:
> Can you use .gitattributes to suppress diffs for .obj files here?

Suppressed in f07b054800686c344c0b8b24417ee81773be9bb3

sloretz committed:
Update README for AMENT_PREFIX_PATH and interactive demo

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz commented:
> LGTM. Mind adding the AMENT_PREFIX_PATH stuff to the README as well?

Rebased and updated the README in 4bd74d1

sloretz commented:
It looks like all the comments have been addressed. Since this is a prototype repo, I'm inclined to merge after giving a day or so for additional reviews.

sloretz commented:
Seemed like there was no opposition to merging in yesterday's meeting, so merging :tada:

EricCousineau-TRI commented:
Sweet, thank you!!!
